pkg/ccl/backupccl/backup_compaction.go

```diff
@@ -0,0 +1,497 @@
+	tasks = append(tasks,
+		func(ctx context.Context) error {
+			return genSpan(ctx, spanCh)
+		})
+	tasks = append(tasks, func(ctx context.Context) error {
+		return runCompaction(ctx, p, encryption, spanCh, newDetails, backupManifest, progCh, store)
+	})
+	tasks = append(tasks, func(ctx context.Context) error {
+		processProg(ctx, progCh, backupManifest)
+		return nil
+	})
+
+	if err := ctxgroup.GoAndWait(ctx, tasks...); err != nil {
+		return err
+	}
+	backupID := uuid.MakeV4()
+	backupManifest.ID = backupID
+
+	if err := backupinfo.WriteBackupManifest(ctx, store, backupbase.BackupManifestName,
+		encryption, kmsEnv, backupManifest); err != nil {
+		return err
+	}
+	if backupinfo.WriteMetadataWithExternalSSTsEnabled.Get(&p.ExecCfg().Settings.SV) {
+		if err := backupinfo.WriteMetadataWithExternalSSTs(ctx, store, encryption,
+			kmsEnv, backupManifest); err != nil {
+			return err
+		}
+	}
+
+	statsTable := getTableStatsForBackup(ctx, p.ExecCfg().TableStatsCache, backupManifest.Descriptors)
+	if err := backupinfo.WriteTableStatistics(ctx, store, encryption, kmsEnv, &statsTable); err != nil {
+		return err
+	}
+
+	if backupinfo.WriteMetadataSST.Get(&p.ExecCfg().Settings.SV) {
+		if err := backupinfo.WriteBackupMetadataSST(ctx, store, encryption, kmsEnv, backupManifest,
+			statsTable.Statistics); err != nil {
+			err = errors.Wrap(err, "writing forward-compat metadata sst")
+			if !build.IsRelease() {
+				return err
+			}
+			log.Warningf(ctx, "%+v", err)
+		}
+	}
+	return nil
+}
+
+func runCompaction(
+	ctx context.Context,
+	p sql.JobExecContext,
+	encryption *jobspb.BackupEncryptionOptions,
+	entries chan execinfrapb.RestoreSpanEntry,
+	details jobspb.BackupDetails,
+	manifest *backuppb.BackupManifest,
+	progCh chan execinfrapb.RemoteProducerMetadata_BulkProcessorProgress,
+	store cloud.ExternalStorage,
+) error {
+	defer close(progCh)
+	var encryptionOptions *kvpb.FileEncryptionOptions
+	if encryption != nil {
+		encryptionOptions = &kvpb.FileEncryptionOptions{Key: encryption.Key}
+	}
+	sinkConf := sstSinkConf{
+		id:       p.ExecCfg().DistSQLSrv.NodeID.SQLInstanceID(),
+		enc:      encryptionOptions,
+		progCh:   progCh,
+		settings: &p.ExecCfg().Settings.SV,
+	}
+	sink := makeFileSSTSink(sinkConf, store, nil)
+	sink.elideMode = manifest.ElidedPrefix
+	defer func() {
+		if err := sink.flush(ctx); err != nil {
+			log.Warningf(ctx, "failed to flush sink: %v", err)
+		}
+		logClose(ctx, sink, "SST sink")
+	}()
+	for {
+		select {
+		case entry, ok := <-entries:
+			if !ok {
+				return nil
+			}
+
+			sstIter, err := openSSTs(ctx, p, entry, encryption, details)
+			if err != nil {
+				return errors.Wrap(err, "opening SSTs")
+			}
+
+			if err := processSpanEntry(ctx, sstIter, sink, manifest); err != nil {
+				return errors.Wrap(err, "processing span entry")
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func processSpanEntry(
+	ctx context.Context, sstIter mergedSST, sink *fileSSTSink, manifest *backuppb.BackupManifest,
```

Comment:
nit: just pass the start and end key, not a pointer to the backup manifest, as you're not modifying anything in it.